### PR TITLE
fix(snapbi): preserve whitespace in JSON string values during minification

### PR DESCRIPTION
**Pull request by salismt**
## Summary
- preserve whitespace inside quoted JSON string values during signature minification
- avoid mutating request payload semantics while still normalizing structural whitespace

### CODE REVIEW

**Comment by salismt** [comment]
@uziwuzzy please review this

File: library/src/test/java/com/midtrans/snapbi/SnapBiWebhookSignatureTest.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    /**
+     * Produces the same signature that Midtrans would produce:
+     * sign(SHA256withRSA, "POST:<urlPath>:<sha256hex(minifiedBody)>:<timestamp>")
+     */
+    private static String signPayload(String notificationPayload, String urlPath, String timeStamp) throws Exception {
+        String minified = SnapBi.minifyJson(notificationPayload);
+        MessageDigest digest = MessageDigest.getInstance("SHA-256");
+        byte[] hash = digest.digest(minified.getBytes());
```
[the remark this answers is not in the document]
Comment (uziwuzzy) [reply library/src/test/java/com/midtrans/snapbi/SnapBiWebhookSignatureTest.java]: agree, so the format is consistent.

File: library/src/main/java/com/midtrans/snapbi/SnapBi.java [hunk 2]
```diff
@@ -294,9 +294,35 @@ public static String getSymmetricSignatureHmacSh512(String accessToken, Map<Stri
         }
     }
 
-    private static String minifyJson(String json) {
-        // Minify JSON by removing whitespace
-        return json.replaceAll("\\s+", "");
+    static String minifyJson(String json) {
```
[the remark this answers is not in the document]
Comment (uziwuzzy) [reply library/src/main/java/com/midtrans/snapbi/SnapBi.java]: agree, this is internal used, not exposed for public.

**Comment by uziwuzzy** [comment]
agree with all the copilot suggestions. I think I would also suggest to create fixtures for the payload, and we should also handle not just VA, since this is being used for all SnapBI features in the library.

 So I would suggest to add fixtures like .json webhook payload for:
  1. VA : pending and settlement (the original bug only showed up on settlement afaik)
  2. QRIS : pending and success
  3. Direct Debit : pending and success

  So that the normal cases and the abnormal/edge cases (spaces in string values, escaped quotes, etc.) across all SnapBI use cases.

File: library/src/test/java/com/midtrans/snapbi/SnapBiWebhookSignatureTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (salismt) [reply library/src/test/java/com/midtrans/snapbi/SnapBiWebhookSignatureTest.java]: Fixed in bdacc8e — uses `StandardCharsets.UTF_8` explicitly.

File: library/src/main/java/com/midtrans/snapbi/SnapBi.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (salismt) [reply library/src/main/java/com/midtrans/snapbi/SnapBi.java]: Fixed in bdacc8e — restored to `private static`. Deleted `SnapBiMinifyJsonTest` (which called the method directly) and moved all coverage into `SnapBiWebhookSignatureTest`, which tests through the public `isWebhookNotificationVerified()` API using JSON fixture files for each payment method.

File: library/src/main/java/com/midtrans/httpclient/error/MidtransError.java [hunk 3]
```diff
@@ -1,19 +1,33 @@
...
 /**
  * MidtransError class to catch error messages
  */
-@Getter
 public class MidtransError extends Exception {
 
     private String message;
     private Integer statusCode;
     private String responseBody;
     private Response response;
```
[the remark this answers is not in the document]
Comment (salismt) [reply library/src/main/java/com/midtrans/httpclient/error/MidtransError.java:14]: Fixed in bdacc8e — added `@Override` annotation.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,11 @@
 ## CHANGELOG MIDTRANS JAVA LIBRARY
 
+## v3.2.2 (April 7, 2026)
+
+Bugs fix:
```
[the remark this answers is not in the document]
Comment (salismt) [reply CHANGELOG.md]: Fixed in bdacc8e — changed to "Bug fixes:".

File: library/src/test/java/com/midtrans/snapbi/SnapBiWebhookSignatureTest.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        String hexHash = bytesToHex(hash).toLowerCase();
```
[the remark this answers is not in the document]
Comment (salismt) [reply library/src/test/java/com/midtrans/snapbi/SnapBiWebhookSignatureTest.java]: Fixed in bdacc8e — removed redundant `.toLowerCase()`.

**Comment by salismt** [comment]
@uziwuzzy updated, please check

**Comment by uziwuzzy** [comment]
LGTM

**Comment by salismt** [comment]
please merged, I cannot merge using this user
